File: reversi/controller.py
import random

LETTERS = 'abcdefgh'
NUMBERS = '12345678'


def convert_to_index(inp):
    y = LETTERS.find(inp[0])
    x = NUMBERS.find(inp[1])
    return y, x


class HumanController():
    def get_input(self, game, color):
        coverage = game.get_coverage(color)
        while True:
            inp = input('> ')
            if inp.lower() == 'restart' or inp.lower() == 'finish':
                return inp.lower()
            if len(inp) < 3 and inp[0] in LETTERS and inp[1] in NUMBERS:
                cell_idx = convert_to_index(inp)
                if cell_idx in list(coverage.keys()):
                    return cell_idx
            print(f'Illegal command: {inp}')


class BotController():
    def get_input(self, game, color):
        coverage = game.get_coverage(color)
        return random.choice(list(coverage.keys()))


class AiController():
    pass


class TestController():
    pass


def prepare():
    player_one = None
    player_two = None
    black_hole = convert_to_index(input())
    ai_color = input()
    if ai_color == "black":
        player_one = AiController()
        player_two = TestController()
    else:
        player_one = TestController()
        player_two - AiController()

    return player_one, player_two, black_hole

File: src/controller.nim
import strformat, tables, sugar, random, sequtils
import reversi

const Letters = "ABCDEFGH"
const Numbers = "12345678"

proc toCellIndex*(inp: string): CellIndex =
    let y = Numbers.find(inp[1])
    let x = Letters.find(inp[0])
    return (y, x)


proc toOutput*(cellIndex: CellIndex): string =
    let y = cellIndex[0]
    let x = cellIndex[1]
    result = fmt"{Letters[x]}{Numbers[y]}"


type Controller* = (Reversi, char) -> string

proc getAiInput(game: Reversi, color: char): string =
    let coverage = game.getCoverage(color)
    if coverage.len() > 0:
        randomize()
        let move = toOutput(sample(toSeq(coverage.keys)))
        echo move
        return move
    return "pass"

proc getOpponentInput(game: Reversi, color: char): string =
    return readLine(stdin)


proc prepare*(): (Controller, Controller, CellIndex) =
    var playerOne, playerTwo: Controller
    let blackHole = toCellIndex(readLine(stdin))
    let aiColor = readLine(stdin)
    if aiColor == "black":
        playerOne = getAiInput
        playerTwo = getOpponentInput
    else:
        playerOne = getOpponentInput
        playerTwo = getAiInput

    return (playerOne, playerTwo, blackHole)
